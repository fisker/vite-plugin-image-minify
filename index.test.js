import url from 'node:url'
import fs from 'node:fs/promises'
import {expect, test} from 'vitest'
import prettyBytes from 'pretty-bytes'
import * as vite from 'vite'
import vitePluginImageSquoosh from './index.js'

const FIXTURES_DIRECTORY = new URL('./fixtures/', import.meta.url)
const prettySize = (size) => `${prettyBytes(size)} (${size})`

async function runVite() {
  /** @type {import('vite').Rollup.RollupOutput} */
  const buildResult = await vite.build({
    root: url.fileURLToPath(FIXTURES_DIRECTORY),
    configFile: false,
    envFile: false,
    write: false,
    logLevel: 'warn',
    plugins: [vitePluginImageSquoosh({__test_enable_cache: false})],
    build: {assetsInlineLimit: 0},
  })

  const assets = buildResult.output
    .filter(
      (assetOrChunk) => assetOrChunk.type === 'asset' && assetOrChunk.name,
    )
    .sort((imageA, imageB) => imageA.name.localeCompare(imageB.name))

  const result = await Promise.all(
    assets.map(async (asset) => {
      const {length: originalSize} = await fs.readFile(
        new URL(asset.name, FIXTURES_DIRECTORY),
      )
      const bundledSize = asset.source.length
      const savedBytes = bundledSize - originalSize
      const savedPercentage = `${((savedBytes / originalSize) * 100).toFixed(2)}%`

      return {
        original: {
          name: asset.name,
          size: prettySize(originalSize),
        },
        bundled: {
          name: asset.fileName,
          size: prettySize(bundledSize),
        },
        saved: {
          percentage: savedPercentage,
          size: prettySize(savedBytes),
        },
      }
    }),
  )

  return result
}

test('Main', async () => {
  const result = await runVite()

  expect(result).toMatchSnapshot()
})
