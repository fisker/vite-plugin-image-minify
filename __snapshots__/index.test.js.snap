// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main 1`] = `
[
  {
    "bundled": {
      "name": "assets/car-lite-okg-C77L.svg",
      "size": "12.9 kB (12907)",
    },
    "original": {
      "name": "car-lite.svg",
      "size": "26.5 kB (26539)",
    },
    "saved": {
      "percentage": "-51.37%",
      "size": "-13.6 kB (-13632)",
    },
  },
  {
    "bundled": {
      "name": "assets/demo-artwork-SS7dJ2TE.jpg",
      "size": "390 kB (390312)",
    },
    "original": {
      "name": "demo-artwork.jpg",
      "size": "2.92 MB (2922539)",
    },
    "saved": {
      "percentage": "-86.64%",
      "size": "-2.53 MB (-2532227)",
    },
  },
  {
    "bundled": {
      "name": "assets/demo-device-screen-CcBXoMsE.png",
      "size": "1.42 MB (1421280)",
    },
    "original": {
      "name": "demo-device-screen.png",
      "size": "1.62 MB (1622142)",
    },
    "saved": {
      "percentage": "-12.38%",
      "size": "-201 kB (-200862)",
    },
  },
  {
    "bundled": {
      "name": "assets/demo-large-photo-DuXcsXKK.jpg",
      "size": "862 kB (862478)",
    },
    "original": {
      "name": "demo-large-photo.jpg",
      "size": "2.79 MB (2790013)",
    },
    "saved": {
      "percentage": "-69.09%",
      "size": "-1.93 MB (-1927535)",
    },
  },
  {
    "bundled": {
      "name": "assets/sofa_grid1x5_420-BqPuAwOk.avif",
      "size": "25.4 kB (25409)",
    },
    "original": {
      "name": "sofa_grid1x5_420.avif",
      "size": "25.4 kB (25409)",
    },
    "saved": {
      "percentage": "0.00%",
      "size": "0 B (0)",
    },
  },
  {
    "bundled": {
      "name": "assets/test-CHyExZGr.jxl",
      "size": "151 B (151)",
    },
    "original": {
      "name": "test.jxl",
      "size": "151 B (151)",
    },
    "saved": {
      "percentage": "0.00%",
      "size": "0 B (0)",
    },
  },
  {
    "bundled": {
      "name": "assets/test-DBME9N5w.webp",
      "size": "2.05 kB (2050)",
    },
    "original": {
      "name": "test.webp",
      "size": "4.88 kB (4880)",
    },
    "saved": {
      "percentage": "-57.99%",
      "size": "-2.83 kB (-2830)",
    },
  },
]
`;
